lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/CachingHdfsInputFile.java

```diff
@@ -0,0 +1,76 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.hdfs;
+
+import alluxio.client.file.CacheContext;
+import alluxio.client.file.URIStatus;
+import alluxio.client.file.cache.CacheManager;
+import alluxio.client.file.cache.LocalCacheFileInStream;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.hadoop.AlluxioHdfsInputStream;
+import alluxio.hadoop.HdfsFileInputStream;
+import alluxio.wire.FileInfo;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoInput;
+import io.trino.hdfs.CallStats;
+import io.trino.hdfs.HdfsContext;
+import io.trino.hdfs.HdfsEnvironment;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class CachingHdfsInputFile
+        extends HdfsInputFile
+{
+    private final CacheManager cacheManager;
+    private final AlluxioConfiguration alluxioConf;
+    private final FileSystem.Statistics statistics = new FileSystem.Statistics("alluxio");
+    private final HashFunction hashFunction = Hashing.murmur3_128();
+
+    public CachingHdfsInputFile(Location location, Long length, Long lastModifiedTime, HdfsEnvironment environment, HdfsContext context, CallStats callStat,
+                                CacheManager cacheManager, AlluxioConfiguration alluxioConf)
+    {
+        super(location, length, lastModifiedTime, environment, context, callStat);
+        this.cacheManager = cacheManager;
+        this.alluxioConf = alluxioConf;
+    }
+
+    @Override
+    public TrinoInput newInput()
+            throws IOException
+    {
+        FileSystem fileSystem = environment.getFileSystem(context, file);
+        FSDataInputStream input = environment.doAs(context.getIdentity(), () -> {
+            String path = file.toString();
+            long lastModifiedTime = lastModified().toEpochMilli();
+
+            FileInfo info = new FileInfo()
+                    .setLastModificationTimeMs(lastModifiedTime)
+                    .setPath(path)
+                    .setFolder(false)
+                    .setLength(length());
+            String cacheIdentifier = hashFunction.hashString(path + lastModifiedTime, UTF_8).toString();
+            URIStatus uriStatus = new URIStatus(info, CacheContext.defaults().setCacheIdentifier(cacheIdentifier));
+            return new FSDataInputStream(new HdfsFileInputStream(
+                    new LocalCacheFileInStream(uriStatus, (uri) -> new AlluxioHdfsInputStream(fileSystem.open(file)), cacheManager, alluxioConf),
```

Comment:
Since we use soft scheduling, this might end up unnecessarily caching data on nodes whenever soft scheduling fails. That is, if a piece of data A is supposed to be cached on node 1, but node 1 is too busy, it might be scheduled to run on node 2. Then node 2 will read A from cloud storage, and add it to its cache. However, the next time data A is supposed to be read, and node 1 is still too busy, we are just as likely to schedule the read to happen on node 3. So we will just end up consuming write capacity and cache capacity from node 2, without really getting any benefit of caching. This will especially be a problem in a cluster with a lot of nodes.

The nodes need a way to know if they are supposed to cache a piece of data or not, and fall back to not caching.

Reply:
That might be possible to mitigate by populating a list of preferred host addresses in the connector split instead of just one host

Reply:
Would this potentially reduce the cache hit rate? For instance, let's say we have 100 nodes, each with a 1 TB cache, a 100 TB of data we want to query, and all data is only cached on a single node. If soft scheduling never fails we will eventually have a 100% cache hit rate. If soft scheduling fails 10% of the time, 10% of the data will be scheduled on a random node not supposed to cache this data, and store "garbage" in its cache. Eventually 10% of all cached data is not cached on a node with the correct cache key. This should give roughly an 80% cache hit rate.

If on the other hand we cache all data on two nodes, then only 50 TB of the data we want to query can be cached. This should give a cache hit rate of 50%, even if soft scheduling never fails.

If nodes only cache data they are supposed to cache, we would get an 90% cache hit rate, if soft scheduling fails 10% of the time.

Reply:
I don't think this is a problem as long as you deterministically provide the same list of preferred host addressees for a given split and the scheduler attempts to schedule splits on these hosts in the provided preferred order. I'm assuming that the probability that all the preferred hosts are too busy to scheduled splits is low with 3 preferred hosts. You might cache some data in multiple nodes but that would be useful when the primary preferred node is too busy.
A couple of alternatives to this are:
1. The worker node checks that it is not the preferred host from the split and uses the non-cached file system implementation when it's not. This would give up on caching if the preferred host is too busy to schedule splits on.
2. The embedded cache implementation itself has the the ability to remotely read cached data from any worker. This way even if a split gets scheduled on a node which didn't cache the data, it can still read the cached data from another node. This is the approach that Rubix takes. However, this is probably not worth the added complexity.